subsys/zbus/zbus.c

```diff
@@ -0,0 +1,254 @@
+#else
+#define _ZBUS_CURRENT_THREAD_INFO "_", k_current_get()
+#endif /* CONFIG_THREAD_NAME */
+
+inline k_timeout_t _zbus_timeout_remainder(uint64_t end_ticks)
+{
+	return K_TICKS((k_ticks_t)MAX(end_ticks - sys_clock_tick_get(), 0));
+}
+
+int zbus_obs_set_enable(struct zbus_observer *obs, bool enabled)
+{
+	if (obs == NULL) {
+		LOG_ERR("Invalid arg, obs is NULL.");
+		_ZBUS_ASSERT_UNREACHABLE;
+		return -EINVAL;
+	}
+	obs->enabled = enabled;
+	return 0;
+}
+
+void __weak _zbus_notify_runtime_listeners(struct zbus_channel *chan)
+{
+}
+
+int __weak _zbus_notify_runtime_subscribers(struct zbus_channel *chan, uint64_t end_ticks)
+{
+	return 0;
+}
+
+void _zbus_notify_listeners(struct zbus_channel *chan)
+{
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->callback != NULL)) {
+			(*obs)->callback(chan);
+		}
+	}
+	_zbus_notify_runtime_listeners(chan);
+}
+
+int _zbus_notify_subscribers(struct zbus_channel *chan, uint64_t end_ticks)
+{
+	int err = 0;
+
+	for (struct zbus_observer **obs = chan->observers; *obs != NULL; ++obs) {
+		if ((*obs)->enabled && ((*obs)->queue != NULL)) {
+			err = k_msgq_put((*obs)->queue, &chan, _zbus_timeout_remainder(end_ticks));
+			if (err) {
+				LOG_ERR("Observer %s queue full during channel %s publishing",
+					(*obs)->name, chan->name);
+				return err;
+			}
+		}
+	}
+	return _zbus_notify_runtime_subscribers(chan, end_ticks);
+}
+
+static inline int zbus_chan_pub_args_check(struct zbus_channel *chan, void *msg,
+					   k_timeout_t timeout)
+{
+	if (k_is_in_isr()) {
+		LOG_ERR("Zbus cannot be used inside a ISR context.");
+		return -EACCES;
+	}
+	if (chan == NULL) {
+		LOG_ERR("Invalid arg, chan is NULL.");
+		return -EINVAL;
+	}
+	if (chan->read_only != 0) {
+		LOG_ERR("Invalid arg, the channel is read-only.");
+		return -EPERM;
+	}
+	if (msg == NULL) {
+		LOG_ERR("Invalid arg, the msg is NULL.");
+		return -EINVAL;
+	}
+	if (chan->validator != NULL) {
+		if (chan->validator(msg, chan->message_size) == false) {
+			LOG_ERR("Invalid arg, the message is not valid.");
+			return -EPERM;
+		}
+	}
+	return 0;
+}
+
+int zbus_chan_pub(struct zbus_channel *chan, void *msg, k_timeout_t timeout)
+{
+	uint64_t end_ticks = sys_clock_timeout_end_calc(timeout);
+	int err = zbus_chan_pub_args_check(chan, msg, timeout);
+
+	if (err) {
+		_ZBUS_ASSERT_UNREACHABLE;
+		return err;
+	}
+
+	err = k_mutex_lock(chan->mutex, _zbus_timeout_remainder(end_ticks));
+	if (err) {
+		LOG_ERR("Could not pub, channel %s is busy.", chan->name);
+		_ZBUS_ASSERT_UNREACHABLE;
+		return err;
+	}
+	memcpy(chan->message, msg, chan->message_size);
+	_zbus_notify_listeners(chan);
+	err = _zbus_notify_subscribers(chan, end_ticks);
+	k_mutex_unlock(chan->mutex);
+	if (err) {
+		LOG_ERR("Could not notify observer its queue is full!");
+		_ZBUS_ASSERT_UNREACHABLE;
+		return err;
+	}
+	LOG_DBG("Channel %s published by thread %s %p", chan->name, _ZBUS_CURRENT_THREAD_INFO);
+	return err;
+}
```

Comment:
I think the whole API can be simplified and use more zephyr provided utilities. Trimming `zbus_chan_pub` could come to something like:

```c
int zbus_chan_pub(struct zbus_channel *chan, void *msg, k_timeout_t timeout)
{
	int ret;
	uint64_t end_ticks = sys_clock_timeout_end_calc(timeout);

	__ASSERT(chan != NULL, "chan is required");
	__ASSERT(msg != NULL, "msg is required");
	__ASSERT(!k_is_in_isr())), "ISR blocked");

	if (chan->read_only) {
		return -EPERM;
	}

	if (chan->validator && !chan->validator(msg, chan->message_size)) {
		return -EACCESS;
	}

	ret = k_mutex_lock(chan->mutex, timeout);
	if (ret) {
		return ret;
	}

	memcpy(chan->message, msg, chan->message_size);

	_zbus_notify_listeners(chan);
	ret = _zbus_notify_subscribers(chan, end_ticks);

	k_mutex_unlock(chan->mutex);

	return ret;
}
```

Reply:
Checking for `msg == NULL` is arguable, if you would add a runtime check or `__ASSERT` as this can be passed along throughout the application code.